# **Exploratory Data Analysis**
---


In [1]:
import numpy as np
import pandas as pd

## **1) Overview**
---
*Dealing with missing values*

### **flights** ([sample](../Scripts/sampling.sql)) & **flights_test**

#### Overview:

##### **flights** sample

In [2]:
flights = pd.read_csv('../Data/files/flights_sample.csv')
flights.head()

M:\Anaconda\envs\LHL\lib\site-packages\IPython\core\interactiveshell.py:3444: DtypeWarning: Columns (25) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


fl_date mkt_unique_carrier branded_code_share mkt_carrier  \
0  2019-05-19                 UA       UA_CODESHARE          UA   
1  2019-05-19                 UA       UA_CODESHARE          UA   
2  2019-05-19                 UA       UA_CODESHARE          UA   
3  2019-05-19                 UA       UA_CODESHARE          UA   
4  2019-05-19                 UA       UA_CODESHARE          UA   

   mkt_carrier_fl_num op_unique_carrier tail_num  op_carrier_fl_num  \
0                4264                EV   N48901               4264   
1                4266                EV   N12540               4266   
2                4272                EV   N11164               4272   
3                4281                EV   N13995               4281   
4                4286                EV   N13903               4286   

   origin_airport_id origin  ... distance  carrier_delay weather_delay  \
0              12266    IAH  ...      127            NaN           NaN   
1              13244    MEM  ...      468            NaN           NaN   
2              12266    IAH  ...     1091            NaN           NaN   
3              11042    CLE  ...      310            NaN           NaN   
4              13061    LRD  ...      301            NaN           NaN   

  nas_delay  security_delay  late_aircraft_delay  first_dep_time  \
0       NaN             NaN                  NaN             NaN   
1       NaN             NaN                  NaN             NaN   
2       NaN             NaN                  NaN             NaN   
3       NaN             NaN                  NaN             NaN   
4       NaN             NaN                  NaN             NaN   

   total_add_gtime  longest_add_gtime  no_name  
0              NaN                NaN      NaN  
1              NaN                NaN      NaN  
2              NaN                NaN      NaN  
3              NaN                NaN      NaN  
4              NaN                NaN      NaN  

[5 rows x 42 columns]

In [3]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2387955 entries, 0 to 2387954
Data columns (total 42 columns):
 #   Column               Dtype  
---  ------               -----  
 0   fl_date              object 
 1   mkt_unique_carrier   object 
 2   branded_code_share   object 
 3   mkt_carrier          object 
 4   mkt_carrier_fl_num   int64  
 5   op_unique_carrier    object 
 6   tail_num             object 
 7   op_carrier_fl_num    int64  
 8   origin_airport_id    int64  
 9   origin               object 
 10  origin_city_name     object 
 11  dest_airport_id      int64  
 12  dest                 object 
 13  dest_city_name       object 
 14  crs_dep_time         int64  
 15  dep_time             float64
 16  dep_delay            float64
 17  taxi_out             float64
 18  wheels_off           float64
 19  wheels_on            float64
 20  taxi_in              float64
 21  crs_arr_time         int64  
 22  arr_time             float64
 23  arr_delay            float64
 24

##### **flights_test**

In [4]:
flights_test = pd.read_csv('../Data/files/flights_test.csv')
flights_test.head()

fl_date mkt_unique_carrier branded_code_share mkt_carrier  \
0  2020-01-01 00:00:00                 WN                 WN          WN   
1  2020-01-01 00:00:00                 WN                 WN          WN   
2  2020-01-01 00:00:00                 WN                 WN          WN   
3  2020-01-01 00:00:00                 WN                 WN          WN   
4  2020-01-01 00:00:00                 WN                 WN          WN   

   mkt_carrier_fl_num op_unique_carrier tail_num  op_carrier_fl_num  \
0                5888                WN   N951WN               5888   
1                6276                WN   N467WN               6276   
2                4598                WN   N7885A               4598   
3                4761                WN   N551WN               4761   
4                5162                WN   N968WN               5162   

   origin_airport_id origin origin_city_name  dest_airport_id dest  \
0              13891    ONT      Ontario, CA            14771  SFO   
1              13891    ONT      Ontario, CA            14771  SFO   
2              13891    ONT      Ontario, CA            14831  SJC   
3              13891    ONT      Ontario, CA            14831  SJC   
4              13891    ONT      Ontario, CA            14831  SJC   

      dest_city_name  crs_dep_time  crs_arr_time dup  crs_elapsed_time  \
0  San Francisco, CA          1810          1945   N                95   
1  San Francisco, CA          1150          1320   N                90   
2       San Jose, CA          2020          2130   N                70   
3       San Jose, CA          1340          1455   N                75   
4       San Jose, CA           915          1035   N                80   

   flights  distance  
0        1       363  
1        1       363  
2        1       333  
3        1       333  
4        1       333

In [5]:
flights_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660556 entries, 0 to 660555
Data columns (total 20 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   fl_date             660556 non-null  object
 1   mkt_unique_carrier  660556 non-null  object
 2   branded_code_share  660556 non-null  object
 3   mkt_carrier         660556 non-null  object
 4   mkt_carrier_fl_num  660556 non-null  int64 
 5   op_unique_carrier   660556 non-null  object
 6   tail_num            659057 non-null  object
 7   op_carrier_fl_num   660556 non-null  int64 
 8   origin_airport_id   660556 non-null  int64 
 9   origin              660556 non-null  object
 10  origin_city_name    660556 non-null  object
 11  dest_airport_id     660556 non-null  int64 
 12  dest                660556 non-null  object
 13  dest_city_name      660556 non-null  object
 14  crs_dep_time        660556 non-null  int64 
 15  crs_arr_time        660556 non-null  int64 
 16  du

#### **flights** sample vs. **flights_test** feature space

To establish the similarities and difference between the feature spaces of **flights** and **flights_test** rather quickly, we can leverage sets to find the intersection and symmetric difference of their features.

In [6]:
flights_initial_features = set(flights.columns)
flights_test_initial_features = set(flights_test.columns)

common_initial_features = flights_initial_features.intersection(flights_test_initial_features)
exclusive_initial_features = flights_initial_features.symmetric_difference(flights_test_initial_features)

Let us check if the flights_test feature space is entirely a subset of flights.

In [7]:
print(
    len(common_initial_features) == len(flights_test_initial_features), 
    len(flights_test_initial_features),
    sep="\n",
    )

True
20


This confirms that the feature space of flights_test is a subset of flights (which is expected given its [description](../Data/descriptions/flights_test_description.md)). Now let us check which additional features we have in the **flights** table.

In [8]:
print('Features exclusive to flights:\n')
for feature in exclusive_initial_features:
    print(feature)

print(f'\nNumber of features exclusive to flights: {len(exclusive_initial_features)}')

Features exclusive to flights:

taxi_out
late_aircraft_delay
diverted
arr_delay
longest_add_gtime
wheels_off
security_delay
cancellation_code
cancelled
dep_time
first_dep_time
arr_time
actual_elapsed_time
nas_delay
no_name
air_time
taxi_in
wheels_on
carrier_delay
total_add_gtime
weather_delay
dep_delay

Number of features exclusive to flights: 22


All of these features exclusive to **flights** are those that we can only know once the plane has already arrived (excluding `no_name`). Therefore, any useful features as part of this set would need to be predicted in order to make an inference about future flight delays.

#### Missing data investigation:

Preview missing data percentages for features with null values.

In [9]:
flights_missing_values = flights.isna().sum() > 0
flights_missing_value_percentage = flights.isna().sum() / flights.shape[0] * 100

flights_missing_value_percentage[flights_missing_values]

tail_num                 0.304947
dep_time                 1.616865
dep_delay                1.647979
taxi_out                 1.707151
wheels_off               1.707151
wheels_on                1.754681
taxi_in                  1.754681
arr_time                 1.718165
arr_delay                1.944132
cancellation_code       98.317305
crs_elapsed_time         0.000168
actual_elapsed_time      1.927172
air_time                 1.963605
carrier_delay           81.124770
weather_delay           81.124770
nas_delay               81.124770
security_delay          81.124770
late_aircraft_delay     81.124770
first_dep_time          99.313597
total_add_gtime         99.313722
longest_add_gtime       99.313681
no_name                100.000000
dtype: float64

In [10]:
flights_test_missing_values = flights_test.isna().sum() > 0
flights_test_missing_value_percentage = flights_test.isna().sum() / flights.shape[0] * 100

flights_test_missing_value_percentage[flights_test_missing_values]

tail_num    0.062773
dtype: float64

##### Empty rows

We will start by ensuring that empty rows are discarded as they offer no information.

In [11]:
flights = flights.dropna(how='all')
flights_test = flights_test.dropna(how='all')

##### `no_name`

no_name is completely blank, so we can drop it with absolutely no information loss.

In [12]:
flights = flights.drop('no_name', axis=1)

##### `cancellation_code`

In [13]:
flights.cancellation_code.unique()

array([nan, 'A', 'C', 'B', 'D'], dtype=object)

It may be worthwhile to investigate the `cancellation_code` feature further, but the missing proportion here should correspond with the `cancelled` feature assuming that codes are only present for cancelled flights (as suggested in the flights description). In other words, most data in this column is ***structually missing*** (i.e. missing because most flights were not cancelled).

In [14]:
flights['cancelled'].value_counts()/flights.shape[0] * 100

0    98.317305
1     1.682695
Name: cancelled, dtype: float64

The percentages support the structurally missing data hypothesis, but let us confirm by inspecting the relevant subset more rigorously.

In [15]:
flights[flights.cancelled > 0]['cancellation_code'].unique()

array(['A', 'C', 'B', 'D'], dtype=object)

In [16]:
flights[flights.cancelled == 0]['cancellation_code'].unique()

array([nan], dtype=object)

We have the most support that we can obtain from the data for the structurally missing hypothesis with respect to `cancellation_code`! It remains possible that human errors (e.g. data entry) are present but obscured in this table.

In [17]:
type(flights['cancellation_code'].unique()[0]) # nan is of type float in this case.

float

In [18]:
# Replace the missing value for now so that we do not consider the column further.
flights['cancellation_code'] = flights['cancellation_code'].apply(lambda code: 'none' if (isinstance(code, float)) else code)
flights['cancellation_code'].unique()

array(['none', 'A', 'C', 'B', 'D'], dtype=object)

##### '`_delay`' columns

All of the delay columns have identical percentages of missing data in **flights** (none are present in **flights_test**). Here we will test the hypothesis that the missing rows for all of these columns are the same. Resultantly, we predict that missing data will be consistent between rows for these features. This may suggest that these data are also ***structurally missing***.  

This may mean that any time a flight is delayed, all delay columns are assigned a baseline value of 0 (as opposed to NaN) and this baseline is incremented appropriately according to the class of delay.

In [19]:
delay_columns = ['carrier_delay', 'weather_delay', 'nas_delay', 'security_delay', 'late_aircraft_delay']
delays_subframe = flights[delay_columns]
delays_subframe.head()
# These 5 rows concurrently displaying NaN provide very weak support for the above hypothesis. Let us investigate more closely.

carrier_delay  weather_delay  nas_delay  security_delay  \
0            NaN            NaN        NaN             NaN   
1            NaN            NaN        NaN             NaN   
2            NaN            NaN        NaN             NaN   
3            NaN            NaN        NaN             NaN   
4            NaN            NaN        NaN             NaN   

   late_aircraft_delay  
0                  NaN  
1                  NaN  
2                  NaN  
3                  NaN  
4                  NaN

Let us check the rows that remain for each delay feature when those with missing values are discarded and ensure that it matches the value predicted given that our hypothesis is correct. Since we know that all delay columns have the same missing value percentage, we can use one as a reference value.

In [20]:
# Determine the actual percentages of non-missing data for the delay rows and the values we expect to see given our hypothesis.
delay_value_percentages = [(delays_subframe[column].dropna().count() / flights.shape[0] * 100)
                           for column in delays_subframe]
predicted_value = (100 - flights_missing_value_percentage.late_aircraft_delay)

np.allclose(delay_value_percentages, predicted_value) # used to prevent rounding errors from causing problems with strict equality

True

It does appear that the missing values correspond with no delay, and are thus structurally missing. It should be noted that the proportion of delayed flights in this sample lines up fairly well with what is observed overall in the publicly available data from the [US Department of Transportation](https://www.transtats.bts.gov/OT_Delay/OT_DelayCause1.asp?20=E). On this basis, we can impute these missing values as 0.

In [21]:
flights[delay_columns] = delays_subframe.fillna(0)

In [22]:
flights_missing_values = flights.isna().sum() > 0
flights_missing_value_percentage = flights.isna().sum() / flights.shape[0] * 100

flights_missing_value_percentage[flights_missing_values]

tail_num                0.304947
dep_time                1.616865
dep_delay               1.647979
taxi_out                1.707151
wheels_off              1.707151
wheels_on               1.754681
taxi_in                 1.754681
arr_time                1.718165
arr_delay               1.944132
crs_elapsed_time        0.000168
actual_elapsed_time     1.927172
air_time                1.963605
first_dep_time         99.313597
total_add_gtime        99.313722
longest_add_gtime      99.313681
dtype: float64

##### `dep_time` and `_gtime` columns

These columns are interesting because not only is the majority of data missing in similar percentages in all 3 of these columns, but the values are not precisely the same. Here we will investigate points of similarity and divergence between observations.

In [23]:
departure_and_gate_time_columns = ['first_dep_time', 'total_add_gtime', 'longest_add_gtime']
departure_and_gate_times = flights[departure_and_gate_time_columns]

In [24]:
some_missing = departure_and_gate_times.dropna(how='all')
no_missing = departure_and_gate_times.dropna()

missing_rows = set(some_missing.index).symmetric_difference(set(no_missing.index))
flights.loc[missing_rows].T # display the transpose for the 3 indices.

813489         2020945                2027091
fl_date                2019-06-20     2018-02-26             2018-02-28
mkt_unique_carrier             B6             AS                     DL
branded_code_share             B6   AS_CODESHARE           DL_CODESHARE
mkt_carrier                    B6             AS                     DL
mkt_carrier_fl_num            631           3296                   5909
op_unique_carrier              B6             KS                     YX
tail_num                   N339JB         N682PA                 N860RW
op_carrier_fl_num             631           3296                   5909
origin_airport_id           10721          10299                  12953
origin                        BOS            ANC                    LGA
origin_city_name       Boston, MA  Anchorage, AK           New York, NY
dest_airport_id             12953          11445                  11298
dest                          LGA            DUT                    DFW
dest_city_name       New York, NY   Unalaska, AK  Dallas/Fort Worth, TX
crs_dep_time                  703           1515                   1755
dep_time                    923.0         1652.0                 1750.0
dep_delay                   140.0           97.0                   -5.0
taxi_out                     32.0           24.0                   18.0
wheels_off                  955.0         1716.0                 1808.0
wheels_on                  1044.0         1945.0                   33.0
taxi_in                       9.0            2.0                    9.0
crs_arr_time                  828           1730                   2121
arr_time                   1053.0         1947.0                   42.0
arr_delay                   145.0          137.0                  201.0
cancelled                       0              0                      0
cancellation_code            none           none                   none
diverted                        0              0                      0
dup                             N              N                      N
crs_elapsed_time             85.0          135.0                  266.0
actual_elapsed_time          90.0          175.0                  472.0
air_time                     49.0          149.0                  445.0
flights                         1              1                      1
distance                      184            792                   1389
carrier_delay               140.0            0.0                    0.0
weather_delay                 0.0           97.0                    0.0
nas_delay                     5.0           40.0                  201.0
security_delay                0.0            0.0                    0.0
late_aircraft_delay           0.0            0.0                    0.0
first_dep_time              653.0         1507.0                 2137.0
total_add_gtime               NaN            NaN                    NaN
longest_add_gtime             NaN            0.0                    NaN

In [25]:
flights.loc[missing_rows]['arr_delay']

813489     145.0
2020945    137.0
2027091    201.0
Name: arr_delay, dtype: float64

There are only 3 rows here with mixed missing values, and all have a departure time. Since none of these flights were cancelled, it is sensible to conclude that either none of them returned to the gate or these may be data entry errors (this is likely for index 2020945, which apparently never left the gate according to the 0 for `longest_add_gtime`).

Overall, because it is not clear what these features are adding and they are missing entirely for the overwhelming majority of the data, we will simply discard them.

In [26]:
flights = flights.drop(columns=departure_and_gate_time_columns)

##### Remainder of missing values:

Given the surplus of observations we have for the **flights** table, and the low missing data percentages for all other observations (<2%), it is worth checking how much data would be lost if all rows missing data were discarded.

In [27]:
total_observations_retained = flights.dropna().shape[0]
percent_data_retained = total_observations_retained / flights.shape[0] * 100

print(f'Total observations retained / {flights.shape[0]}: {total_observations_retained}')
print(f'Percentage retained: {percent_data_retained:.2f}')

Total observations retained / 2387955: 2339962
Percentage retained: 97.99


Given the minimal percentage loss in tandem with the massive sample size retained by dropping rows with missing values, the tradeoff it worth it in this instance.

In [28]:
flights = flights.dropna()

By the above logic, the tradeoff should be similarly valuable for **flights_test**:

In [29]:
total_observations_retained = flights_test.dropna().shape[0]
percent_data_retained = total_observations_retained / flights_test.shape[0] * 100

print(f'Total observations retained / {flights_test.shape[0]}: {total_observations_retained}')
print(f'Percentage retained: {percent_data_retained:.2f}')

Total observations retained / 660556: 659057
Percentage retained: 99.77


In [30]:
flights_test = flights_test.dropna()

### **passengers**

#### Overview:

In [31]:
passengers = pd.read_csv('../Data/files/passengers.csv')
passengers.head()

departures_scheduled  departures_performed  payload  seats  passengers  \
0                     2                     2    75800    300         244   
1                     2                     2    72400    256         219   
2                     2                     2    76000    300         240   
3                     2                     2   104600    376         187   
4                     2                     2    72700    256         214   

   freight  mail  distance  ramp_to_ramp  air_time  ... dest_country  \
0        0     0      1020           310       259  ...           US   
1      458   950       255           140        90  ...           US   
2       75   335       601           208       174  ...           US   
3        0    58       304           156       103  ...           US   
4        3   553       507           196       156  ...           US   

   dest_country_name aircraft_group aircraft_type aircraft_config month  year  \
0      United States              6           694               1  2017     3   
1      United States              6           698               1  2017     3   
2      United States              6           694               1  2017     3   
3      United States              6           622               1  2017     3   
4      United States              6           698               1  2017     3   

   distance_group  class  data_source  
0               3      F           DU  
1               1      F           DU  
2               2      F           DU  
3               1      F           DU  
4               2      F           DU  

[5 rows x 38 columns]

In [32]:
passengers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2350497 entries, 0 to 2350496
Data columns (total 38 columns):
 #   Column                 Dtype 
---  ------                 ----- 
 0   departures_scheduled   int64 
 1   departures_performed   int64 
 2   payload                int64 
 3   seats                  int64 
 4   passengers             int64 
 5   freight                int64 
 6   mail                   int64 
 7   distance               int64 
 8   ramp_to_ramp           int64 
 9   air_time               int64 
 10  unique_carrier         object
 11  airline_id             int64 
 12  unique_carrier_name    object
 13  region                 object
 14  carrier                object
 15  carrier_name           object
 16  carrier_group          int64 
 17  carrier_group_new      int64 
 18  origin_airport_id      int64 
 19  origin_city_market_id  int64 
 20  origin                 object
 21  origin_city_name       object
 22  origin_country         object
 23  origin_

Interestingly, this data comes from a number of different sources. While we do not necessarily require this column for future analysis, my curiosity will not allow me to ignore this.

In [33]:
passengers.data_source.unique()

array(['DU', 'IU', 'IF', 'DF'], dtype=object)

There are 4 unique sources. I wonder what their distribution looks like...

In [34]:
passengers.data_source.value_counts() / passengers.shape[0] * 100

DU    80.252772
IU    10.931901
IF     8.170825
DF     0.644502
Name: data_source, dtype: float64

It looks as though the data is heavily skewed towards the `DU` source at more than 80%. `IU` and `IF` are similarly represented, and `DF` is by far the least frequently observed.

#### Missing data investigation:

In [35]:
passengers_missing_values = passengers.isna().sum() > 0
passengers_missing_value_percentage = passengers.isna().sum() / passengers.shape[0] * 100

passengers_missing_value_percentage[passengers_missing_values]

origin_country    0.000128
dtype: float64

In [37]:
total_observations_retained = passengers.dropna().shape[0]
percent_data_retained = total_observations_retained / passengers.shape[0] * 100

print(f'Total observations retained / {passengers.shape[0]}: {total_observations_retained}')
print(f'Percentage retained: {percent_data_retained:.4f}')

Total observations retained / 2350497: 2350494
Percentage retained: 99.9999


In [38]:
passengers = passengers.dropna()

### **fuel_consumption**

#### Overview:

In [39]:
fuel = pd.read_csv('../Data/files/fuel_consumption.csv')
fuel.head()

month  airline_id unique_carrier carrier  \
0      1         NaN            NaN     0JQ   
1      1     21352.0            0WQ     0WQ   
2      1     21645.0            23Q     23Q   
3      1     21652.0            27Q     27Q   
4      1     20408.0             5V      5V   

                                        carrier_name  carrier_group_new  \
0                                                NaN                  1   
1                                  Avjet Corporation                  1   
2                              Songbird Airways Inc.                  1   
3                 Jet Aviation Flight Services, Inc.                  1   
4  Tatonduk Outfitters Limited d/b/a Everts Air A...                  1   

   sdomt_gallons  satl_gallons  spac_gallons  slat_gallons  ...  sdomt_cost  \
0            0.0           0.0           0.0           0.0  ...           0   
1            0.0           0.0           0.0           0.0  ...           0   
2            0.0           0.0           0.0           0.0  ...           0   
3            0.0           0.0           0.0           0.0  ...           0   
4       260848.0           0.0           0.0           0.0  ...      522405   

   satl_cost  spac_cost  slat_cost  sint_cost  ts_cost  tdomt_cost  tint_cost  \
0        0.0        0.0        0.0        0.0        0        5100        0.0   
1        0.0        0.0        0.0        0.0        0      396216   140239.0   
2        0.0        0.0        0.0        0.0        0           0        0.0   
3        0.0        0.0        0.0        0.0        0           0        0.0   
4        0.0        0.0        0.0        0.0   522405      569497        0.0   

   total_cost  year  
0        5100  2016  
1      536455  2016  
2           0  2016  
3           0  2016  
4      569497  2016  

[5 rows x 25 columns]

In [40]:
fuel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3035 entries, 0 to 3034
Data columns (total 25 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   month              3035 non-null   int64  
 1   airline_id         3034 non-null   float64
 2   unique_carrier     3034 non-null   object 
 3   carrier            3035 non-null   object 
 4   carrier_name       3034 non-null   object 
 5   carrier_group_new  3035 non-null   int64  
 6   sdomt_gallons      3034 non-null   float64
 7   satl_gallons       3034 non-null   float64
 8   spac_gallons       3034 non-null   float64
 9   slat_gallons       3033 non-null   float64
 10  sint_gallons       3034 non-null   float64
 11  ts_gallons         3034 non-null   float64
 12  tdomt_gallons      3032 non-null   float64
 13  tint_gallons       3034 non-null   float64
 14  total_gallons      3029 non-null   float64
 15  sdomt_cost         3035 non-null   int64  
 16  satl_cost          3034 

#### Missing data investigation:

In [41]:
fuel_missing_values = fuel.isna().sum() > 0
fuel_missing_value_percentage = fuel.isna().sum() / fuel.shape[0] * 100

fuel_missing_value_percentage[fuel_missing_values]

airline_id        0.032949
unique_carrier    0.032949
carrier_name      0.032949
sdomt_gallons     0.032949
satl_gallons      0.032949
spac_gallons      0.032949
slat_gallons      0.065898
sint_gallons      0.032949
ts_gallons        0.032949
tdomt_gallons     0.098847
tint_gallons      0.032949
total_gallons     0.197694
satl_cost         0.032949
spac_cost         0.032949
slat_cost         0.032949
sint_cost         0.032949
tint_cost         0.032949
dtype: float64

In [44]:
total_observations_retained = fuel.dropna().shape[0]
percent_data_retained = total_observations_retained / fuel.shape[0] * 100

print(f'Total observations retained / {fuel.shape[0]}: {total_observations_retained}')
print(f'Percentage retained: {percent_data_retained:.2f}')

Total observations retained / 3035: 3025
Percentage retained: 99.67


In [45]:
fuel = fuel.dropna()